consensus/src/data_manager.rs

```diff
@@ -0,0 +1,234 @@
+
+            let _ = self
+                .quorum_store_wrapper_tx
+                .load()
+                .as_ref()
+                .unwrap()
+                .as_ref()
+                .clone()
+                .try_send(WrapperCommand::CleanRequest(logical_time, digests));
+
+            if !payload_is_empty {
+                let expired_set = self.expiration_status.lock().expire(logical_time.round());
+                for expired in expired_set {
+                    self.digest_status.remove(&expired);
+                }
+            }
+        }
+    }
+
+    pub async fn update_payload(&self, block: &Block) {
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            if block.payload().is_some() {
+                match block.payload().unwrap() {
+                    Payload::InQuorumStore(proofs) => {
+                        if !self.digest_status.contains_key(&block.id()) {
+                            let receivers = self
+                                .request_data(
+                                    proofs.clone(),
+                                    LogicalTime::new(block.epoch(), block.round()),
+                                )
+                                .await;
+                            self.digest_status
+                                .insert(block.id(), DataStatus::Requested(receivers));
+                            self.expiration_status
+                                .lock()
+                                .add_item(block.id(), block.round());
+                        }
+                    }
+                    Payload::Empty => {}
+                    Payload::DirectMempool(_) => {
+                        unreachable!()
+                    }
+                }
+            }
+        }
+    }
+
+    pub async fn get_data(&self, block: &Block) -> Result<Vec<SignedTransaction>, Error> {
+        if block.payload().is_none() || block.payload().unwrap().is_empty() {
+            return Ok(Vec::new());
+        }
+        if self.quorum_store_enabled.load(Ordering::Relaxed) {
+            if let Payload::InQuorumStore(proofs) = block.payload().unwrap() {
+
+                // let data_status = self.digest_status.entry(block.id());
+                match self.digest_status.entry(block.id()) {
+                    dashmap::mapref::entry::Entry::Occupied(mut entry) => match entry.get_mut() {
+                        DataStatus::Cached(data) => {
+                            return Ok(data.clone());
+                        }
+                        DataStatus::Requested(receivers) => {
+                            let mut vec_ret = Vec::new();
+                            debug!("QSE: waiting for data on {} receivers", receivers.len());
+                            for rx in receivers {
+                                match rx.await {
+                                    Err(_) => {
+                                        debug!("Oneshot channel to get a batch was dropped");
```

Comment:
What happens here? if we just print a debug message and continue, could return a wrong list of transactions?
Is timeout Ok(Err(e))? 
Basically what's confusing is why isn't this either an invariant violation, or the same flow as timeout (retry)

Reply:
Good point. In this case, we are probably still in an old epoch.
We return an error now, hopefully triggering a state sync.